homeassistant/components/xiaomi_miio/sensor.py

```diff
@@ -145,21 +174,112 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
                         for variable in sensor_variables
                     ]
                 )
-
-    if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+    elif config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
         host = config_entry.data[CONF_HOST]
         token = config_entry.data[CONF_TOKEN]
-        name = config_entry.title
-        unique_id = config_entry.unique_id
-
-        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+        model = config_entry.data[CONF_MODEL]
+        device = None
+        if model in MODELS_HUMIDIFIER_MIOT:
+            device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+            coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+            attributes = HUMIDIFIER_ATTRIBUTES_MIOT
+        elif model.startswith("zhimi.humidifier."):
+            device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+            coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+            attributes = HUMIDIFIER_ATTRIBUTES
+        if device:
+            for attribute in attributes:
+                entities.append(
```

Comment:
This structure doesn't look very nice. A nicer way would be to have a "device" always available that contains either an empty list of sensors ("attributes") and handle this in a generic manner.

There will be much more sensors exposed by other platforms in the future, so it would be great to get this done properly to make it simple to convert other platforms away from state attributes.

Reply:
I do not want risking with touching the AirQualityMonitor to much in this PR. This could be updated later too. For now I have made adding sensor generic and took it out of the if then construct.